### ISSUE TITLE
OSLC Query BNF for wildcards seems questionable

### ISSUE THREAD

**Issue by jamsden** [jamsden/oslc-core-test#23]
I was about to answer this question [https://jazz.net/forum/questions/227587/how-to-use-rm-apioslc-where-clause-using-not-equal-null-conditions](https://jazz.net/forum/questions/227587/how-to-use-rm-apioslc-where-clause-using-not-equal-null-conditions) with the link to the OSLC Query spec but then I saw !="\\\*" on the linked docs in [https://www.ibm.com/support/knowledgecenter/SSZRHJ/com.ibm.mif.doc/gp\\\_intfrmwk/oslc/r\\\_oslc\\\_query\\\_params.html](https://www.ibm.com/support/knowledgecenter/SSZRHJ/com.ibm.mif.doc/gp\_intfrmwk/oslc/r\_oslc\_query\_params.html).

That made me read BNF extra carefully and from the shortened subset below you can see my concern:

oslc\\\_where ::= "oslc.where=" compound\\\_term  
compound\\\_term ::= simple\\\_term (space? boolean\\\_op space? simple\\\_term)\\\*  
simple\\\_term ::= term | scoped\\\_term  
space ::= " " /\\\* a space character \\\*/  
boolean\\\_op ::= "and"  
term ::= identifier\\\_wc comparison\\\_op value | identifier\\\_wc space in\\\_op space? in\\\_val  
scoped\\\_term ::= identifier\\\_wc "

{" compound\\\_term "}

"  
identifier\\\_wc ::= identifier | wildcard

This query is valid:

oslc.where=\\\* > 10

While this one is invalid:

oslc.where=name=\\\*

Now that I look at my examples, this actually makes sense (leaving aside a question why would you filter by any property being greater than a certain number), while IBM manual‘s "The parent!="\\\*" query is semantically equivalent to the parent="NULL" query." does not!

Do you have any automatic BNF validators to check the sanity of the BNF further?

**Comment by DavidJHoney** [comment]
I agree that we should make the wildcard for both oslc.where and oslc.select optional. That needs to be described in a normative section and, I suggest, as a comment in the formal syntax.

If the syntax of oslc.select is required to be the same as oslc.properties, then we should mostly reference the syntax of oslc.properties in OSLC Core. However, because of potential constraints specific to query, I think we need to keep the section on oslc.select in the query spec.

**Comment by DavidJHoney** [comment]
Another use of wildcard is in nested queries. For example:

*   { foaf:name = "username" }

This is even more problematic for implementations using RDB persistence. Again, there does not appear to be any way for a client to discover whether this wildcard use case is supported by an implementation.

**Comment by berezovskyi** [comment]
Thanks David, by "we" here I was speaking as a Core TC member, not as a Lyo lead ![](https://issues.oasis-open.org/images/icons/emoticons/smile.png)

In Lyo we indeed use ANTLR grammar, but it was created by hand and resembles the BNF in the spec rather loosely:

oslc\\\_where : compound\\\_term  
;

compound\\\_term : simple\\\_term ( boolean\\\_op simple\\\_term)\\\* -> ^( ‘terms‘ simple\\\_term (simple\\\_term)\\\* )  
;

simple\\\_term : term | scoped\\\_term  
;

space : ‘ ‘! ;

boolean\\\_op : ‘ and ‘! ;

term : term\\\_simple | in\\\_term ;

term\\\_simple  
: identifier\\\_wc comparison\\\_op value -> ^( ‘simple\\\_term‘ identifier\\\_wc comparison\\\_op value )  
;

in\\\_term : identifier\\\_wc in\\\_op space? in\\\_val -> ^( ‘in\\\_term‘ identifier\\\_wc in\\\_val )  
;

scoped\\\_term : identifier\\\_wc ‘

{‘ compound\\\_term ‘}

‘ -> ^( ‘compound\\\_term‘ identifier\\\_wc compound\\\_term )  
;

identifier\\\_wc : identifier | wildcard ;

identifier : prefixedName ;

wildcard : ASTERISK ;

comparison\\\_op : ‘=‘ | ‘!=‘ | ‘<‘ | ‘>‘ | ‘<=‘ | ‘>=‘ ;

in\\\_op : ‘ in‘ ;

in\\\_val : ‘\[‘ value (‘,‘ value)\\\* ‘\]‘ -> ^( ‘in\\\_values‘ value (value)\\\* )  
;

value : iriRef | literal\\\_value ;

literal\\\_value : boolean\\\_val | decimal | string\\\_esc | typed\\\_string | langed\\\_string  
;

typed\\\_string  
: string\\\_esc ( ‘^^‘ prefixedName ) -> ^( ‘typed\\\_value‘ string\\\_esc prefixedName? )  
;

langed\\\_string  
: string\\\_esc LANGTAG -> ^( ‘langed\\\_value‘ string\\\_esc LANGTAG )  
;

boolean\\\_val : BOOLEAN ;

decimal : DECIMAL ;

string\\\_esc : STRING\\\_LITERAL ;

properties :  
property (‘,‘ property )\\\*  
;

property :  
prefixedName ( nested\\\_property )\\\*  
;

nested\\\_property :  
‘

{‘ property (‘,‘ property )\\\* ‘}

‘  
;

iriRef  
: IRI\\\_REF

prefixedName  
;

prefixedName  
: PNAME\\\_LN

PNAME\\\_NS  
;

\\\*\\\*\\\*

On the MAY implementation of the wildcard, I agree that we can make it if there are no major OSLC 2 query implementation that allow queries like \\\*

{ foaf:name = "username" }

. Regarding SPARQL operator compabiity, I, unfortunatly, agree with the original spec that mandates the equivance. The reason is simple: it has rather simple semantics defined in XML Query terms and are not that hard to implement. Also, OSLC was supposed to provide a sane way for enterprises to move towards RDF and, hence, RDF stores (all of which support SPARQL nowadays).

My suggestions:

*   publish OSLC Query BNF as the stripped-down ANTLR source (which uses EBNF), so that we can check it in Antlr IDE
*   make wildcard property handling a MAY

**Comment by DavidJHoney** [comment]
Eclipse Lyo defines the grammar using Antlr. I did something similar for the Global Configuration Management application‘s implementation of OSLC Query using an updated Antlr grammar.

The other thing the spec should probably say is that using a wildcard for a property might not be supported by an implementation. For RDF persistence and SPARQL, it‘s easy to specify a graph match and filter. But for RDB persistence with SQL, then the implementation is awkward, involving potential many terms to separately query multiple columns and potentially multiple tables. Hence an implementation might choose to report wildcard property names as an error.

The spec currently does not provide any advice about how a client discovers such limitations.

**Comment by berezovskyi** [comment]
BTW, we declare that OSLC Query is defined in EBNF, but it‘s not! EBNF notation does not use ::=, requires semicolons etc. The "EBNF" definition fails the check via [http://www.icosaedro.it/bnf\\\_chk/bnf\\\_chk-on-line.html](http://www.icosaedro.it/bnf\_chk/bnf\_chk-on-line.html)
